buildSrc/subprojects/parser/src/main/java/org/apache/calcite/buildtools/parser/DialectGenerate.java

```diff
@@ -0,0 +1,419 @@
+    }
+  };
+
+  private final File dialectDirectory;
+  private final File rootDirectory;
+  private final String outputFile;
+
+  public DialectGenerate(File dialectDirectory, File rootDirectory, String outputFile) {
+    this.dialectDirectory = dialectDirectory;
+    this.rootDirectory = rootDirectory;
+    this.outputFile = outputFile;
+  }
+
+  /**
+   * Extracts functions and prints the results for the given dialect.
+   */
+  public void run() {
+    Map<String, String> functions = extractFunctions();
+    // TODO(AndrewPochapsky): Remove this once generation logic added.
+    for (Map.Entry<String, String> entry : functions.entrySet()) {
+      System.out.println(entry.getKey() + "=" + entry.getValue() + "\n");
+    }
+  }
+
+  /**
+   * Traverses the parsing directory structure and extracts all of the
+   * functions located in *.ftl files into a Map.
+   */
+  public Map<String, String> extractFunctions() {
+    Map<String, String> functionMap = new LinkedHashMap<String, String>();
+    traverse(getTraversalPath(rootDirectory), rootDirectory, functionMap);
+    return functionMap;
+  }
+
+  /**
+   * Generates the parserImpls.ftl file for the dialect.
+   *
+   * @param functions The functions to place into the output file
+   */
+  public void generateParserImpls(Map<String, String> functions) {
+    // TODO(AndrewPochapsky): Add generation logic.
+  }
+
+  /**
+   * Gets the traversal path for the dialect by "subtracting" the root
+   * absolute path from the dialect directory absolute path.
+   *
+   * @param rootDirectoryFile The file for the root parsing directory
+   */
+  private Queue<String> getTraversalPath(File rootDirectoryFile) {
+    Path dialectPath = dialectDirectory.toPath();
+    Path rootPath = rootDirectory.toPath();
+    dialectPath = dialectPath.subpath(rootPath.getNameCount(), dialectPath.getNameCount());
+    Queue<String> pathElements = new LinkedList<>();
+    dialectPath.forEach(p -> pathElements.add(p.toString()));
+    return pathElements;
+  }
+
+  /**
+   * Traverses the determined path given by the directories queue. Once the queue is
+   * empty, the dialect directory has been reached. In that case any *.ftl
+   * file should be processed and no further traversal should happen.
+   *
+   * @param directories The directories to traverse in topdown order
+   * @param currentDirectory The current directory the function is processing
+   * @param functionMap The map to which the parsing functions will be added to
+   */
+  private void traverse(
+      Queue<String> directories,
+      File currentDirectory,
+      Map<String, String> functionMap) {
+    File[] files = currentDirectory.listFiles();
+    // Ensures that files are processed first.
+    Arrays.sort(files, fileComparator);
+    String nextDirectory = directories.peek();
+    for (File f : files) {
+      String extension = "";
+      String fileName = f.getName();
+      int i = fileName.lastIndexOf('.');
```

Comment:
I think lines 141-145 could be simplified by using fileName.endsWith(".ftl")

Reply:
Good point, will change.